File: frontend/src/lib/processPrefill.ts
/**
 * Vorbelegung von Feldern aus den Daten der angemeldeten Person (Spiegel von
 * services/process_prefill.apply_prefill). Nur für die Anzeige beim Anlegen –
 * autoritativ setzt der Server die Werte erneut.
 */
import type { FieldDef } from '@/types/process'

type Values = Record<string, unknown>

function resolvePath(obj: unknown, path: string): unknown {
  let cur: unknown = obj
  for (const part of path.split('.')) {
    if (cur && typeof cur === 'object') cur = (cur as Record<string, unknown>)[part]
    else return undefined
  }
  return cur
}

/** `profile` = Antwort von /auth/profile (employee + Konto-Felder wie phone/email). */
export function applyPrefill(
  fields: FieldDef[], profile: Record<string, unknown> | null, values: Values,
): Values {
  const out: Values = { ...values }
  const employee = (profile?.employee as Record<string, unknown>) ?? {}
  for (const f of fields) {
    if (!f.prefill?.field) continue
    const src = f.prefill.source === 'user' ? (profile ?? {}) : employee
    const val = resolvePath(src, f.prefill.field)
    // Nur Skalare: eine nicht aufgelöste Relation (Objekt/Array) darf nicht als
    // Wert landen (sonst „[object Object]"). Pfad auf ein Unterfeld zeigen lassen.
    if (val !== undefined && val !== null && val !== '' && typeof val !== 'object') {
      out[f.key] = val
    }
  }
  return out
}

File: backend/services/process_prefill.py
"""Felder beim Anlegen aus den Daten der angemeldeten Person vorbelegen (§ prefill).

Für read-only Antragsteller-Felder: der Wert kommt aus dem Directus-Mitarbeiter-
Datensatz (user["employee"]) bzw. den Session-Feldern und wird serverseitig
AUTORITATIV gesetzt (überschreibt Client-Eingaben – manipulationssicher). Wird nur
beim ANLEGEN angewandt (Momentaufnahme, wer den Antrag gestellt hat); spätere
Änderungen der Stammdaten wirken nicht rückwirkend.
"""
from typing import Any, Optional

from backend.schemas.process_definition import ProcessDefinition


def _resolve_path(obj: Any, path: str) -> Optional[Any]:
    """dot-Pfad in einem (verschachtelten) dict auflösen; Relationen wie
    „location.name". Listen werden nicht durchlaufen."""
    cur = obj
    for part in path.split("."):
        if isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur


def apply_prefill(defn: ProcessDefinition, values: dict, user: dict) -> dict:
    """Felder mit `prefill`-Spec aus den Daten der angemeldeten Person füllen.

    source "employee" → user["employee"] (Directus-Stammdaten), source "user" →
    das Session-User-Objekt. Ein leerer/fehlender Quellwert lässt das Feld
    unangetastet (kein Überschreiben mit None).
    """
    out = dict(values)
    employee = user.get("employee") or {}
    for f in defn.fields:
        pf = f.prefill
        if not pf:
            continue
        src = employee if pf.source == "employee" else user
        val = _resolve_path(src, pf.field)
        if val is not None and val != "":
            out[f.key] = val
    return out
